lib/srv/sess.go

```diff
@@ -1238,43 +1252,44 @@ func (s *session) startTerminal(ctx context.Context, scx *ServerContext) error {
 	return nil
 }
 
-// newRecorder creates a new events.StreamWriter to be used as the recorder
+// newRecorder creates a new [events.SessionSetterRecorder] to be used as the recorder
 // of the passed in session.
-func newRecorder(s *session, ctx *ServerContext) (events.StreamWriter, error) {
+func newRecorder(s *session, ctx *ServerContext) (events.SessionSetterRecorder, error) {
 	// Nodes discard events in cases when proxies are already recording them.
 	if s.registry.Srv.Component() == teleport.ComponentNode &&
 		services.IsRecordAtProxy(ctx.SessionRecordingConfig.GetMode()) {
-		return &events.DiscardStream{}, nil
+		return events.WithNoOpSetter(events.NewDiscardRecorder()), nil
 	}
 
-	streamer, err := s.newStreamer(ctx)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	rec, err := events.NewAuditWriter(events.AuditWriterConfig{
-		// Audit stream is using server context, not session context,
-		// to make sure that session is uploaded even after it is closed
-		Context:      ctx.srv.Context(),
-		Streamer:     streamer,
-		SessionID:    s.id,
-		Clock:        s.registry.clock,
-		Namespace:    s.serverMeta.ServerNamespace,
-		ServerID:     s.serverMeta.ServerID,
-		RecordOutput: ctx.SessionRecordingConfig.GetMode() != types.RecordOff,
-		Component:    teleport.Component(teleport.ComponentSession, ctx.srv.Component()),
-		ClusterName:  ctx.ClusterName,
-	})
+	rec, err := recorder.New(
+		events.SetterConfig{
+			SessionID:   s.id,
+			ServerID:    s.serverMeta.ServerID,
+			Namespace:   s.serverMeta.ServerNamespace,
+			ClusterName: ctx.ClusterName,
+		},
+		events.SessionWriterConfig{
+			// Audit stream is using server context, not session context,
+			// to make sure that session is uploaded even after it is closed
+			Context:   ctx.srv.Context(),
+			SessionID: s.id,
+			Clock:     s.registry.clock,
+			Component: teleport.Component(teleport.ComponentSession, ctx.srv.Component()),
+		},
+		ctx.SessionRecordingConfig,
+		filepath.Join(
+			ctx.srv.GetDataDir(), teleport.LogsDir, teleport.ComponentUpload,
+			events.StreamingSessionsDir, ctx.srv.GetNamespace(),
+		),
```

Comment:
This logic for building the directory path for a recording seems to be duplicated in a few places - what do you think about a utility in `lib/events` to consolidate this?

Reply:
I rolled that into `lib/events/recorder.New` instead and simplified how `recorder.New` is called